DOCSP-45330: inheritance (#66)
mongodb/docs-mongoid: source/data-modeling/inheritance.txt

   :values: reference

.. meta::
   :keywords: ruby framework, odm, relationship, code example, polymorphic

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to implement **inheritance** into your
{+odm+} models. Inheritance allows you to apply the characteristics of
one "parent" class to one or more "child" classes.

{+odm+} supports inheritance in top level and embedded documents.
When a child model class inherits from a parent class, {+odm+} copies
the parent class's fields, associations, validations, and scopes to
the child class.

Assign Inheritance
------------------

When creating a child model class, use the ``<`` character to implement
inheritance from a specified parent class. The following model classes
demonstrate how to create parent and child classes between the
``Person``, ``Employee``, and ``Manager`` models:

.. literalinclude:: /includes/data-modeling/inheritance.rb
   :start-after: start-simple-inheritance
   :end-before: end-simple-inheritance
   :language: ruby
   :emphasize-lines: 7, 14
   :dedent:

When you perform data operations by using the preceding models,

> instances of ``Person``, ``Employee``, or ``Manager`` are all saved in the
> ``people`` collection. {+odm+} sets the ``_type`` discriminator field to

Reviewer: [s] Suggestion for active voice:
```suggestion
{+odm} saves instances of ``Person``, ``Employee``, or ``Manager`` in the
``people`` collection. {+odm+} sets the ``_type`` discriminator field to
```

the model class name in documents to ensure that documents are returned
as the expected types when you retrieve them.

Embedded Documents
~~~~~~~~~~~~~~~~~~

You can also implement an inheritance pattern in embedded associations.
Similar to the behavior of top-level model classes, {+odm+} sets the
``_type`` discriminator field in embedded documents depending on the
model class used to create them.

The following example adds an embedded association to the ``Person``
model and creates parent and child models for the embedded ``Info``
class:

.. literalinclude:: /includes/data-modeling/inheritance.rb
   :start-after: start-embedded-inheritance
   :end-before: end-embedded-inheritance
   :language: ruby
   :emphasize-lines: 5, 14, 17, 22
   :dedent:

Query Behavior
~~~~~~~~~~~~~~

When you query on a child model class, the query returns only documents

> in which the value of the ``_type`` field match the queried class or

Reviewer: ```suggestion
in which the value of the ``_type`` field matches the queried class or
```

You can set a global key by adding the following code to your
application *before* defining any model classes:

.. code-block:: ruby

   Mongoid.discriminator_key = "sub_type"

All classes use ``sub_type`` as the discriminator key and do not include
the ``_type`` field.

.. note::

   You must set the discriminator key on the global level before defining
   any child classes for the classes to use that global value. If you set
   the global key after defining child classes, your saved documents
   contain the default ``_type`` field.

Change the Discriminator Value
------------------------------

You can customize the value that {+odm+} sets as the discriminator value
in MongoDB. Use the ``discriminator_value()`` method when defining a
class to customize the discriminator value, as shown in the following
example:

.. code-block:: ruby
   :emphasize-lines: 6

   class Employee
     include Mongoid::Document
   
     field :company, type: String

     self.discriminator_value = "Worker"
   end

When you create an instance of ``Employee``, the document's ``_type``
discriminator field has a value of ``"Worker"`` instead of the class
name.

> .. note::
> 
>    Because the discriminator value customization is declared in child classes,
>    you must load the child classes retrieved by a query *before* sending
>    that query. In the preceding example, the ``Employee`` class definition
>    must be loaded before you query on ``Person`` if the returned documents could
>    potentially be instances of ``Employee``. Autoloading isn't able to resolve
>    the discriminator value ``"Worker"`` to return the document as an
>    instance of ``Employee``.

Reviewer: [s] Can potentially omit extra words for brevity and make this note easier to follow, for example:
```suggestion
.. note::

   Because the discriminator value customization is declared in child classes,
   you must load the child classes retrieved by a query *before* sending
   that query. In the preceding example, the ``Employee`` class definition
   must be loaded before you query on ``Person`` if the returned documents could
   be instances of ``Employee``. Autoloading cannot resolve
   the discriminator value ``"Worker"`` to return the document as an
   instance of ``Employee``.
```
